README.md
# AWS (Amazon Web Services) library for Ruby

This gem provides access to many AWS cloud services.
Unlike many other AWS libaries this gem is a very thin adaptation layer over the AWS APIs.
It is highly meta-programmed and exposes the exact AWS API calls with the
exact AWS parameters. There are two big benefits to this approach: you don't have to translate betwen the
AWS docs and the library in order to figure out what to do, and there's nothing in the AWS APIs you can't
call through the library. The library doesn't even need to be updated when AWS introduces new API features.
The downside to this approach is that when things don't work it's sometimes more difficult to figure out
what's happening... Another downside is that method names and parameter names follow the AWS API spec and
thereby run against Ruby conventions.

If you encounter problems please open a github issue.

## Docs

For some getting-started info see further down but for detailed docs see
http://rightscale.github.io/right_aws_api/frames.html#!file.README.html

## Features

The gem supports the following AWS services out of the box:

Web Services:

- {RightScale::CloudApi::AWS::AS::Manager      Auto Scaling (AS)}
- {RightScale::CloudApi::AWS::CF::Manager      Cloud Front (CF)}
- {RightScale::CloudApi::AWS::CFM::Manager     Cloud Formation (CFM)}
- {RightScale::CloudApi::AWS::CW::Manager      Cloud Watch (CW)}
- {RightScale::CloudApi::AWS::EB::Manager      Elastic Beanstalk (EB)}
- {RightScale::CloudApi::AWS::EC::Manager      Elasti Cache (EC)}
- {RightScale::CloudApi::AWS::EC2::Manager     Elastic Compute Cloud (EC2)}
- {RightScale::CloudApi::AWS::ELB::Manager     Elastic Load Balancing (ELB)}
- {RightScale::CloudApi::AWS::EMR::Manager     Elastic Map Reduce (EMR)}
- {RightScale::CloudApi::AWS::IAM::Manager     Identity and Access Management (IAM)}
- {RightScale::CloudApi::AWS::RDS::Manager     Relational Database Service (RDS)}
- {RightScale::CloudApi::AWS::Route53::Manager Route 53 (Route53)}
- {RightScale::CloudApi::AWS::Route53::Domain::Manager Route 53 Domain (Route53Domain)}
- {RightScale::CloudApi::AWS::S3::Manager      Simple Storage Service (S3)}
- {RightScale::CloudApi::AWS::SDB::Manager     Simple DB (SDB)}
- {RightScale::CloudApi::AWS::SNS::Manager     Simple Notification Service (SNS)}
- {RightScale::CloudApi::AWS::SQS::Manager     Simple Queue Service (SQS)}

Product Advertising API:

- {RightScale::CloudApi::ECS::PA::Manager      Product Advertising API (PA)}

Market Web Services:

- {RightScale::CloudApi::MWS::CartInformation::Manager              Cart Information}
- {RightScale::CloudApi::MWS::CustomerInformation::Manager          Customer Information}
- {RightScale::CloudApi::MWS::Feeds::Manager                        Feeds}
- {RightScale::CloudApi::MWS::Finances::Manager                     Finances}
- {RightScale::CloudApi::MWS::FulfillmentInboundShipment::Manager   Fulfillment Inbound Shipment}
- {RightScale::CloudApi::MWS::FulfillmentInventory::Manager         Fulfillment Inventory}
- {RightScale::CloudApi::MWS::FulfillmentOutboundShipment::Manager  Fulfillment Outbound Shipment}
- {RightScale::CloudApi::MWS::Orders::Manager                       Orders}
- {RightScale::CloudApi::MWS::Products::Manager                     Products}
- {RightScale::CloudApi::MWS::Recommendations::Manager              Recommendations}
- {RightScale::CloudApi::MWS::Reports::Manager                      Reports}
- {RightScale::CloudApi::MWS::Sellers::Manager                      Sellers}
- {RightScale::CloudApi::MWS::Subscriptions::Manager                Subscriptions}
- {RightScale::CloudApi::MWS::Webstore::Manager                     Webstore}

And it is easy to add support for other. You will need to refer to
the AWS docs (http://aws.amazon.com/documentation/) for all the API params and usage explanations.

## Basic usage

### Amazon Elastic Compute Cloud (EC2)

This library supports all existing and future EC2 API calls. If you know that EC2 supports a
call and you know what params it accepts - just call the method with those params.

#### Basics

```ruby
  require "right_aws_api"

  key      = ENV['AWS_ACCESS_KEY_ID']
  secret   = ENV['AWS_SECRET_ACCESS_KEY']
  endpoint = 'https://ec2.us-east-1.amazonaws.com'
  ec2      = RightScale::CloudApi::AWS::EC2::Manager.new(key, secret, endpoint)

  ec2.ThisCallMustBeSupportedByEc2('Param.1' => 'A', 'Param.2' => 'B')
```

#### EC2 Instances

```ruby
  require "right_aws_api"

  ec2 = RightScale::CloudApi::AWS::EC2::Manager.new(key, secret, endpoint)

  # Get a list of your instances
  ec2.DescribeInstances

  # Describe custom Instance(s)
  ec2.DescribeInstances('InstanceId' => "i-2ba7c640")
  ec2.DescribeInstances('InstanceId' => ["i-2ba7c640", "i-7db9101e"])

  # Describe Instances with filteringSame (another way, the result is the same):
  ec2.DescribeInstances(
    'Filter' => [
      {'Name'  => 'architecture',
       'Value' => 'i386'},
      {'Name'  => 'availability-zone',
       'Value' => [ 'us-east-1a', 'us-east-1d' ]},
      {'Name'  => 'instance-type',
       'Value' =>  'm1.small'} ]
  )

  # Run an new instance:
  ec2.RunInstances(
    'ImageId'            => 'ami-8ef607e7',
    'MinCount'           => 1,
    'MaxCount'           => 1,
    'KeyName'            => 'kd: alex',
    'UserData'           => RightScale::CloudApi::Utils::base64en('Hehehehe!!!!'),
    'InstanceType'       => 'c1.medium',
    'ClientToken'        => RightScale::CloudApi::Utils::generate_token,
    'SecurityGroupId'    => ['sg-f71a089e', 'sg-c71a08ae' ],
    'Placement+'         => {
       'AvailabilityZone' => 'us-east-1d',
       'Tenancy'          => 'default' },
    'BlockDeviceMapping' => [
      {'DeviceName' => '/dev/sdc',
       'Ebs'        => {
         'SnapshotId'          => 'snap-e40fd188',
         'VolumeSize'          => 3,
         'DeleteOnTermination' => true}} ]
  )

  # Terminate your instance:
  ec2.TerminateInstances("InstanceId" => ["i-67c87504", "i-67c87504"])
```

### Amazon Simple Storage Service (S3)

```ruby
  require "right_aws_api"

  key      = ENV['AWS_ACCESS_KEY_ID']
  secret   = ENV['AWS_SECRET_ACCESS_KEY']
  endpoint = 'https://s3.amazonaws.com'
  s3       = RightScale::CloudApi::AWS::S3::Manager::new(key, secret, endpoint)

  s3.ListBuckets
  s3.GetBucketAcl('Bucket' => 'my-lovely-bucket')
  s3.GetObject('Bucket' => 'my-lovely-bucket', 'Object' => 'fairies.txt')
```

### Amazon Simple Queue Service (SQS)

```ruby
  require "right_aws_api"

  key            = ENV['AWS_ACCESS_KEY_ID']
  secret         = ENV['AWS_SECRET_ACCESS_KEY']
  account_number = ENV['AWS_ACCOUNT_NUMBER']
  endpoint       = 'https://sqs.us-east-1.amazonaws.com'
  sqs            = RightScale::CloudApi::AWS::SQS::Manager.new(key, secret, account_number, endpoint)

  # List all queues
  sqs.ListQueues

  # Create a new one
  sqs.CreateQueue(
    'QueueName' => 'myCoolQueue',
    'Attribute' => [
       { 'Name'  => 'VisibilityTimeout',  'Value' => 40 },
       { 'Name'  => 'MaximumMessageSize', 'Value' => 2048 } ])

  # Send a message
  message = URI::escape('Woohoo!!!')
  sqs.SendMessage('myCoolQueue', 'MessageBody' => message)

  # Receive a message
  sqs.ReceiveMessage('myCoolQueue')

  # Kill the queue
  sqs.DeleteQueue('myCoolQueue')
```

### Amazon Product Advertising API (PA)

```ruby

  require "right_aws_api"

  key      = ENV['AWS_ACCESS_KEY_ID']
  secret   = ENV['AWS_SECRET_ACCESS_KEY']
  endpoint = 'https://webservices.amazon.com'
  paa      = RightScale::CloudApi::ECS::PA::Manager.new(key, secret, endpoint)

  paa.ItemLookup(
    'AssociateTag'  => 'weird-tag',
    'IdType'        => 'ASIN',
    'ItemId'        => 'B00TRAO8HK',
    'ResponseGroup' => 'OfferSummary'
  )
```

### Amazon Market Web Services API (MWS)

```ruby

  key      = ENV['AWS_ACCESS_KEY_ID']
  secret   = ENV['AWS_SECRET_ACCESS_KEY']
  endpoint = 'https://mws.amazonservices.com'

  # Webstore
  ws = RightScale::CloudApi::MWS::Webstore::Manager.new(key, secret, endpoint)
  ws.GetServiceStatus

  # Reports
  rd = RightScale::CloudApi::MWS::Reports::Manager.new(key, secret, endpoint)
  rd.GetReportList(
    'SellerId'      => 'foo',
    'MarketplaceId' => 'bar',
  )

  # Recommendations
  rd = RightScale::CloudApi::MWS::Recommendations::Manager.new(key, secret, endpoint)
  rd.ListRecommendations(
    'SellerId'      => 'foo',
    'MarketplaceId' => 'bar',
  )
```

### Options

There is a way to provide extra options when you instantiate a new manager:

```ruby
  options = { :key => value }
  ec2 = RightScale::CloudApi::AWS::EC2::Manager.new(key, secret, endpoint, options)
  sqs = RightScale::CloudApi::AWS::SQS::Manager.new(key, secret, account_number, endpoint, options)
  s3  = RightScale::CloudApi::AWS::S3::Manager::new(key, secret, endpoint, options)
  etc
```

The options allow you to tweak the managers behavoir. Here is a list of the options that make sense for AWS services:

Name                     | Type    | Default                     | Description
-------------------------| --------| ----------------------------|--------------
:abort_on_timeout        | Boolean | false                       | When set to +true+ the gem does not perform a retry call when there is a connection timeout. This may help you to deal with request idempotence issue. Lets say you make a create call and you get back a timeout. It is possible that AWS created a new resource but just failed to report it properly. It is better to stop here rather than keep retrying creating more and more resources.
:api_version             | String  | see service Manager         | The required cloud API version if it is different from the default one.
:cache                   | Boolean | false                       | Cache cloud responses when possible so that we don't parse them again if cloud response does not change (see cloud specific ApiManager definition).
:cloud                   | Hash    | {}                          | A set of cloud or service specific options. There is an only option for AWS S3 so far: :no_dns_buckets => false/true.
:connection_open_timeout | Integer | up to NetHttpPersistent gem | Connection open timeout (in seconds).
:connection_read_timeout | Integer | up to NetHttpPersistent gem | Connection read timeout (in seconds).
:connection_retry_count  | Integer | 3                           | Max number of establish connection retry attempts before it gives up.
:connection_retry_delay  | Float   | 0.5                         | Initial retry backoff delay in seconds. The value is doubled on every retry attempt.
:connection_verify_mode  | Integer | OpenSSL::SSL::VERIFY_PEER   | Try OpenSSL::SSL::VERIFY_NONE is there is an SSL sertificate issue with th eremote end. This may happen when working with DNS-line S3 buckets.
:headers                 | Hash    | {}                          | A set of request headers to be added to every API request.
:logger                  | Logger  | -> STDOUT                   | Current logger. When nil is given it logs to '/dev/nul'.
:log_filter_patterns     | Array   | see DEFAULT_LOG_FILTERS     | A set of log filters that define what to log (see {RightScale::CloudApi::CloudApiLogger}).
:params                  | Hash    | {}                          | A set of URL params to be added to every API request.
:raw_response            | Boolean | false                       | By default the gem parses all XML and JSON responses and returns them as ruby Hashes. Sometimes it is not what one would want (Amazon S3 GetObject for example). Setting this option to +true+ forces the gem to return the body of the response as it is.

For more options see https://github.com/rightscale/right_cloud_api_base/blob/master/lib/base/api_manager.rb

### Response

It is easy to get the last HTTP request and response data:

```ruby
  s3.PutObject('Bucket' => 'a.b.c.d.1.com', 'Object' => '13', :body => 'hahaha') #=> ''

  s3.request.verb    # => 'put'
  s3.request.path    # => '/13'
  s3.request.headers # =>
    {
      "content-type"         => ["binary/octet-stream"],
      "content-length"       => [6],
      "content-md5"          => ["EBpuyfk4iF3wpE8gRY0utA=="],
      "x-amz-content-sha256" => ["23453452345234523452345"],
      "x-amz-date"           => ["20150123T225348Z"],
      "x-amz-expires"        => [3600],
      "host"                 => ["a.b.c.d.1.com.s3.amazonaws.com"],
      "authorization"        => ["AWS4-HMAC-SHA256 Credential=000/20150123/us-east-1/s3/aws4_request, SignedHeaders=content-length;content-md5;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-expires, Signature=111"]}
  s3.request.body    #=> "hahaha"

  s3.response.code    #=> '200'
  s3.response.headers #=>
    {
      "x-amz-id-2"       => ["wB/XOQ+dfgdfgdfgwgfdg"],
      "x-amz-request-id" => ["FEC7DEE7C51ACAB3"],
      "date"             => ["Fri, 23 Jan 2015 22:53:49 GMT"],
      "etag"             => ["\"101a6ec9f938885df0a44f20458d2eb4\""],
      "content-length"   => ["0"],
      "server"           => ["AmazonS3"]}
  s3.response.body    #=> ''

```

Furthermore every response value may return its HTTP response code and HTTP headers through 'metadata' method:

```ruby
response1 = s3.GetObject('Bucket' => 'a.b.c.d.1.com', 'Object' => '13')
response2 = s3.GetObject('Bucket' => 'a.b.c.d.1.com', 'Object' => '14')

response1 #=> "hahaha"
response1.metadata #=>
  { :code    => '200',
    :headers =>
    {"x-amz-id-2"       => ["sdasfdasd="],
     "x-amz-request-id" => ["B2AB70B2D081BF2B"],
     "date"             => ["Fri, 23 Jan 2015 23:02:36 GMT"],
     "last-modified"    => ["Fri, 23 Jan 2015 22:53:49 GMT"],
     "etag"             => ["\"2345435234523452345324\""],
     "accept-ranges"    => ["bytes"],
     "content-type"     => ["binary/octet-stream"],
     "content-length"   => ["6"],
     "server"           => ["AmazonS3"]}}

response2 #=> "hohohohoho"
response2.metadata #=>
{ :code    => '200',
  :headers =>
  {"x-amz-id-2"       => ["asdasdf="],
   "x-amz-request-id" => ["50DBC360C934C548"],
   "date"             => ["Fri, 23 Jan 2015 23:02:36 GMT"],
   "last-modified"    => ["Fri, 23 Jan 2015 23:01:55 GMT"],
   "etag"             => ["\"2345345243523452345234\""],
   "accept-ranges"    => ["bytes"],
   "content-type"     => ["binary/octet-stream"],
   "content-length"   => ["10"],
   "server"           => ["AmazonS3"]}}
```



## Dependencies

This gem depends on a base gem which is shared across all RightScale cloud libraries:
https://github.com/rightscale/right_cloud_api_base

### (c) 2014 by RightScale, Inc., see the LICENSE file for the open-source license.


HISTORY
== current (not released as a gem yet)
== 2016-10-18, v0.3.5
  - Fixed a bug in MWS error parsing
  - Bumped right_cloud_api_base version to not use net-http-persistent 3.0+
== 2015-11-11, v0.3.4
  - Add LimitRateExceeded to the retryable error types
== 2015-09-24, v0.3.3
  - Add iam regex for AWS China
== ?? , v0.3.2
  - Fixed:
    - Route53
      - All POST actions were broken because of the missing 'xmlns' attribute.
  - Added new services:
    - Product Advertising API
    - Market Web Service API:
      - Cart Information
      - Customer Information
      - Feeds
      - Finances
      - Fulfillment Inbound Shipment
      - Fulfillment Inventory
      - Fulfillment Outbound Shipment
      - Orders
      - Products
      - Recommendations
      - Reports
      - Sellers
      - Subscriptions
      - Webstore

== 2015-06-02, v0.3.1
  - AWS china support
== 2015-03-12, v0.3.0
  - Fixed bugs in Route53:
    - new param added to ChangeResourceRecordSets: HostedZoneId -> TargetHostedZoneId
    - method renamed: CreateHostedZones -> CreateHostedZone
  - Route53 API 2013-04-01 support added
  - Route53 Domain API v20140515 support added
  - Switched to the newer version of right_cloud_api_base gem

== 2015-01-16, v0.2.3
  - Added an ability to not use DNS like buckets in AWS S3
  - Added support for S3 folders creation

== 2015-01-16, v0.2.2
  - Some minor bugs were fixed in singature V4

== 2014-10-30, v0.2.1
  - Added support for AWS v4 signature and make it the default

== 2013-06-28, pre-release


lib/cloud/aws/s3/manager.rb
#--
# Copyright (c) 2013 RightScale, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# 'Software'), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#++

require "cloud/aws/base/helpers/utils"
require "cloud/aws/s3/parsers/response_error"
require "cloud/aws/s3/wrappers/default"
require "cloud/aws/s3/routines/request_signer"

module RightScale
  module CloudApi
    module AWS

      # Simple Storage Service namespace
      #
      # @api public
      #
      module S3

        # Amazon Simple Storage Service (S3) compatible manager (thread safe).
        #
        # There are 2 ways of using S3 API manager: _HTTP verb calls_ or _helper methods_
        #
        # ### HTTP verbs
        #
        # The manager supports following HTTP verbs: get, post, put, head and delete,
        # and all of them share the same syntax:
        #
        # ```ruby
        #  s3.post(path, :params => Hash, :headers => Hash, :body => 'foo-bar', :options => Hash)
        # ```
        #
        # ### Helper methods
        #
        # Eventhough the _HTTP verbs_ are powerfull they are not too handy.
        # If you like to call API actions by their name you may find our helper methods usefull.
        #
        # In most cases these methods use the same names and they take the same parameters
        # that Amazon uses in their docs.
        #
        # You can find use case examples for both the verbs and the methods below.
        # The helper method definitions can be found here: https://github.com/rightscale/right_aws_api/blob/master/lib/cloud/aws/s3/wrappers/default.rb
        #
        # @example
        #  require "right_aws_api"
        #
        #  s3 = RightScale::CloudApi::AWS::S3::Manager.new(key, secret, 'https://s3.amazonaws.com')
        #
        # @example
        #  # -- Using HTTP verbs --
        #
        #  # List all buckets
        #  s3.get
        #
        #  # List bucket objects
        #  s3.get('devs-us-east')
        #
        #  # Get bucket ACL
        #  s3.get('devs-us-east', :params => {'acl' => nil} )
        #
        #  # Put bucket ACL
        #  s3.put('devs-us-east',
        #         :params  => {'acl' => nil},
        #         :body    => access_control_policy_xml,
        #         :headers => { 'content-type' => 'application/xml' }
        #  )
        #
        #  # Get bucket Versions
        #  s3.get('devs-us-east', :params => {'version' => nil} )
        #
        #  # Get object
        #  s3.get('devs-us-east/boot1.jpg')
        #  # Do not parse response if Amazon reports back XML or JSON content-type
        #  s3.get('devs-us-east/boot1.xml', :options => { :raw_response => true})
        #
        #  # Get object, force set content-type
        #  s3.get('kd-kd-kd-1/boot1.jpg', :params => { 'response-content-type' => 'image/jpeg'})
        #
        #  # Put object
        #  # Do not forget to provide a proper 'content-type' header because the default
        #  # one is set to 'binary/octet-stream'
        #  s3.put('devs-us-east/boot1.jpg',
        #          :body    => 'This is my object DATA. WooHoo!!!',
        #          :headers => {'content-type' => 'text/plain'})
        #
        #  # Create a folder
        #  s3.put('devs-us-east/logs/')
        #
        # @example
        #   # A simple example of a multi-thread file download:
        #   threads    = []
        #   file_size  = 3257230
        #   chunks     = 3
        #   chunk_size = file_size / chunks
        #   chunks.times do |i|
        #     from_byte = i * chunk_size
        #     to_byte   = from_byte + chunk_size - 1
        #     to_byte  += file_size % chunks if i + 1 == chunks
        #     threads << Thread::new {
        #       Thread.current[:my_file] = s3.get('devs-us-east/xxx/boot.jpg', {:headers => {'Range' => "bytes=#{from_byte}-#{to_byte}"}})
        #     }
        #   end
        #   file_body = ''
        #   threads.each do |thread|
        #     thread.join
        #     file_body << thread[:my_file]
        #   end
        #   file_body.size #=> 3257230
        #
        #
        # @example
        #   # Download into IO object
        #   File.open('/tmp/boot.jpg','w') do |file|
        #     s3.get('devs-us-east/kd/boot.jpg') do |chunk|
        #       file.write(chunk)
        #     end
        #   end
        #
        # @example
        #  # -- Using helper methods --
        #
        #  # List all buckets
        #  s3.ListAllMyBuckets
        #
        #  # List bucket objects
        #  s3.ListObjects('Bucket' => 'devs-us-east')
        #
        #  # Get bucket ACL
        #  s3.GetBucketAcl('Bucket' => 'devs-us-east')
        #
        #  # Get bucket Versions
        #  s3.GetBucketVersions('Bucket' => 'devs-us-east')
        #
        #  # Get object
        #  s3.GetObject('Bucket' => 'devs-us-east', 'Object' => 'boot1.jpg')
        #
        #  # Get object, force set content-type
        #  s3.GetObject('Bucket' => 'devs-us-east', 'Object' => 'boot1.jpg',
        #               :params => { 'response-content-type' => 'image/jpeg'})
        #
        #  # Put object
        #  # P.S. 'content-type' is 'binary/octet-stream' by default
        #  s3.PutObject('Bucket' => 'devs-us-east',
        #               'Object' => 'boot1.jpg',
        #               :body    => file_content,
        #               :headers => {'content-type' => 'image/jpeg'})
        #
        #  # Create a folder
        #  s3.PutObject('Bucket' => 'devs-us-east', 'Object' => 'logs/', :body => '')
        #
        # @example
        #
        #  # List all buckets
        #  s3.ListAllMyBuckets #=>
        #    {"ListAllMyBucketsResult"=>
        #      {"Buckets"=>
        #        {"Bucket"=>
        #          [{"Name"=>"CI_right_test",
        #            "CreationDate"=>"2011-05-25T20:46:28.000Z"},
        #           {"Name"=>"CR_right_test",
        #            "CreationDate"=>"2011-06-08T20:46:32.000Z"},
        #           {"Name"=>"DarrylTest",
        #            "CreationDate"=>"2011-06-03T03:43:08.000Z"},
        #           {"Name"=>"RightScalePeter",
        #            "CreationDate"=>"2008-10-28T03:59:20.000Z"}]},
        #       "Owner"=>
        #        {"DisplayName"=>"fghsfg",
        #         "ID"=>"16144ab2929314cc309ffe736daa2b264357476c7fea6efb2c3347ac3ab2792a"},
        #       "@xmlns"=>"http://s3.amazonaws.com/doc/2006-03-01/"}}
        #
        #  # List bucket objects
        #  s3.ListObjects('Bucket' => 'devs-us-east', 'max-keys' => 3, 'prefix' => 'kd') #=>
        #    {"ListBucketResult"=>
        #      {"MaxKeys"=>"3",
        #       "IsTruncated"=>"false",
        #       "Name"=>"devs-us-east",
        #       "Marker"=>nil,
        #       "Contents"=>
        #        {"LastModified"=>"2010-08-26T12:23:30.000Z",
        #         "StorageClass"=>"STANDARD",
        #         "ETag"=>"\"3c9a2717e34efedb6d6ac007b2acb8df\"",
        #         "Owner"=>
        #          {"DisplayName"=>"thve",
        #           "ID"=>
        #            "16144ab2929314cc309ffe736daa2b264357476c7fea6efb2c3347ac3ab2792a"},
        #         "Size"=>"3257230",
        #         "Key"=>"kd/boot.jpg"},
        #       "@xmlns"=>"http://s3.amazonaws.com/doc/2006-03-01/",
        #       "Prefix"=>"kd"}}
        #
        #
        # @example
        #  # Get
        #  s3.GetBucketCors('Bucket' => 'my-bucket' ) #=>
        #    {"CORSConfiguration"=>
        #        {"@xmlns"=>"http://s3.amazonaws.com/doc/2006-03-01/",
        #        "CORSRule"=>
        #          [{"AllowedOrigin"=>"http://www.example.com",
        #            "AllowedMethod"=>["PUT", "POST"],
        #            "MaxAgeSeconds"=>"2000",
        #            "ExposeHeader"=>"x-amz-server-side-encryption"},
        #          {"AllowedOrigin"=>"*", "AllowedMethod"=>"GET", "MaxAgeSeconds"=>"2001"}]}}
        #
        #
        # @example
        #  # Put
        #  cors_rules = [
        #    {'AllowedOrigin' => 'http://www.example.com',
        #     'AllowedMethod' => ['PUT', 'POST'],
        #     'MaxAgeSeconds' => 3000,
        #     'ExposeHeader'  => 'x-amz-server-side-encryption' },
        #    {'AllowedOrigin' => '*',
        #     'AllowedMethod' => 'GET',
        #     'MaxAgeSeconds' => 3000 } ]
        #  s3.PutBucketCors('Bucket' => 'kd-ver-test', 'CORSRule' => cors_rules ) #=> ''
        #
        #  # .. or
        #  body = "<CORSConfiguration><CORSRule><AllowedOrigin>http://www.example.com</AllowedOrigin>"+
        #         "<AllowedMethod>PUT</AllowedMethod><AllowedMethod>POST</AllowedMethod>"+
        #         "<MaxAgeSeconds>3000</MaxAgeSeconds>..</CORSConfiguration>"
        #  s3.PutBucketCors('Bucket' => 'my-bucket', :body => body ) #=> ''
        #
        #
        # @example
        #  # Delete
        #  s3.DeleteBucketCors('Bucket' => 'my-bucket' ) #=> ''
        #
        #
        # @example
        #  # Bucket Tagging
        #  # Get
        #  s3.GetBucketTagging('Bucket' => 'my-bucket' ) #=>
        #    {"Tagging"=> {
        #       "TagSet"=> {
        #          "Tag"=>[
        #            {"Key"=>"Project",
        #              "Value"=>"Project One"},
        #             {"Key"=>"User",
        #              "Value"=>"jsmith"}]}}}
        #
        #
        # @example
        #  # Delete
        #  s3.DeleteBucketTagging('Bucket' => 'my-bucket' ) #=> ''
        #
        #
        # @example
        #  # Put
        #  tagging_rules = [
        #    {"Key"=>"Project",
        #     "Value"=>"Project One"},
        #    {"Key"=>"User",
        #     "Value"=>"jsmith"} ]
        #  s3.PutBucketTagging('Bucket' => 'my-bucket', 'TagSet' => tagging_rules ) #=> ''
        #
        #
        # @example
        #  # Bucket Lifecycle
        #  # Get
        #  s3.GetBucketLifecycle('Bucket' => 'my-bucket' ) #=>
        #    {"LifecycleConfiguration"=> {
        #       "Rule"=>[{
        #         "ID" => "30-day-log-deletion-rule",
        #         "Prefix" => "logs",
        #         "Status" => "Enabled",
        #         "Expiration" => { "Days" => 30 }}]}}
        #
        #
        # @example
        #  # Delete
        #  s3.DeleteBucketLifecycle('Bucket' => 'my-bucket' ) #=> ''
        #
        #
        # @example
        #  # Put
        #  lifecycle_rules = [
        #    { "ID"         => "30-day-log-deletion-rule",
        #      "Prefix"     => "logs",
        #      "Status"     => "Enabled",
        #      "Expiration" => {"Days" => 30}},
        #    {"ID"          => "delete-documents-rule",
        #      "Prefix"     => "documents",
        #      "Status"     => "Enabled",
        #      "Expiration" => { "Days" => 365 }}]
        #  s3.PutBucketLifecycle('Bucket' => 'my-bucket', 'Rule' => lifecycle_rules ) #=> ''
        #
        #
        # @example
        #   link = RightScale::CloudApi::AWS::S3::Link::Manager::new(key, secret, endpoint)
        #   link.get(
        #     'devs-us-east/kd/Константин',
        #     :params => { 'response-content-type' => 'image/peg'}
        #   ) #=>
        #     'https://devs-us-east.s3.amazonaws.com/kd%2F%D0%9A%D0%BE%D0%BD%D1%81%D1%82%D0%B0%
        #      D0%BD%D1%82%D0%B8%D0%BD?AWSAccessKeyId=AK...TA&Expires=1436557118&
        #      Signature=hg...%3D&response-content-type=image%2Fpeg'
        #
        # @example
        #   link.ListAllMyBuckets #=>
        #     'https://s3.amazonaws.com/?AWSAccessKeyId=AK...TA&Expires=1436651780&
        #      Signature=XK...53s%3D'
        #
        # @example
        #   link = RightScale::CloudApi::AWS::S3::Link::Manager::new(key, secret, endpoint)
        #   link.GetObject('Bucket' => 'foo', 'Object' => 'bar') #=>
        #     'https://foo.s3.amazonaws.com/bar?AWSAccessKeyId=AK...TA&Expires=1436557118&
        #      Signature=hg...%3D&response-content-type=image%2Fpeg'
        #
        # @example
        #   # Do not use DNS-like bucket hosts but put buckets into path
        #   link = RightScale::CloudApi::AWS::S3::Link::Manager::new(key, secret, endpoint, :cloud => { :no_dns_buckets => true })
        #   link.GetObject('Bucket' => 'foo', 'Object' => 'bar') #=>
        #     'https://s3.amazonaws.com/foo/bar?AWSAccessKeyId=AK...TA&Expires=1436557118&
        #      Signature=hg...%3D&response-content-type=image%2Fpeg'
        #
        # @see ApiManager
        # @see Wrapper::DEFAULT.extended Wrapper::DEFAULT.extended (click [View source])
        # @see http://docs.aws.amazon.com/AmazonS3/latest/API/APIRest.html
        #
        class Manager < CloudApi::Manager
        end


        # Amazon Simple Storage Service (S3) compatible manager (thread unsafe).
        #
        # @see Manager
        #
        class  ApiManager < CloudApi::ApiManager

          # S3 Error
          class Error < CloudApi::Error
          end

          include Mixin::QueryApiPatterns

          set_routine CloudApi::RetryManager
          set_routine CloudApi::RequestInitializer
          set_routine AWS::S3::RequestSigner
          set_routine CloudApi::RequestGenerator
          set_routine CloudApi::ConnectionProxy
          set_routine CloudApi::ResponseAnalyzer
          set_routine CloudApi::ResponseParser
          set_routine CloudApi::ResultWrapper

          set :response_error_parser => Parser::AWS::S3::ResponseError


          # Constructor
          #
          # @param [String] aws_access_key_id
          # @param [String] aws_secret_access_key
          # @param [String] endpoint
          # @param [Hash]   options
          #
          # @example
          #   # see Manager class
          #
          # @see Manager
          #
          def initialize(aws_access_key_id, aws_secret_access_key, endpoint, options={})
            credentials = { :aws_access_key_id     => aws_access_key_id,
                            :aws_secret_access_key => aws_secret_access_key }
            super(credentials, endpoint, options)
          end


          # Makes an API call to AWS::S3 compatible cloud
          #
          # @param [String,Symbol] verb  'get' | 'put' | etc
          # @param [Objects] args
          #
          # @return [Object]
          #
          # @example
          #   api(verb,                  opts={})
          #   api(verb, 'bucket',        opts={})
          #   # Where opts may have next keys: :options, :headers, :params, :body
          #   api(verb, 'bucket/object', opts={})
          #
          def api(verb, *args, &block)
            relative_path = args.first.is_a?(String) ? args.shift : ''
            opts          = args.shift || {}
            raise Error::new("Opts must be Hash not #{opts.class.name}") unless opts.is_a?(Hash)
            process_api_request(verb, relative_path, opts, &block)
          end
        end

      end
    end
  end
end
